### PR TITLE
refactor: Implement typed `SelectionInfo` object

### PR DESCRIPTION
**Pull request by mrvinceit**
I took a look at what was done with the SelectionRelations and thought about how best to introduce a typed object, that works for both relations and complex objects.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
     "apollo-server-types": "3.7.1",
     "class-validator": "0.14.0",
     "clsx": "^1.2.1",
+    "graphql-parse-resolve-info": "^4.12.3",
```
Comment (TriPSs) [inline package.json:30]: Not used?

**Review by TriPSs — COMMENTED** [review]
Think it's better the keep the return types of the different decorators separate. Reason for keeping them separate was so that only users who enable the `enableLookAhead` will have this, and there will be no performance setback for the users that don't need / want it. 

Users that want the info van then overwrite the resolver, get the info and pass whatever they want down them self to the query service.

Could you also add a example JSON of the output? Also have the feeling that the parser is a lot more complex than the initial one. It's also missing the parsing of the args.

File: package.json (hunk 1 above)
Comment (mrvinceit) [reply package.json:30]: It is used in [packages/query-graphql/src/decorators/graphql-resolve-info.utils.ts](https://github.com/TriPSs/nestjs-query/pull/71/files#diff-a8c4f481736b2c380dd5f16a492f5efd1da1e0e7a5f9f82d3408f1a1de71479f)  in `createSelectionInfo`

Comment (mrvinceit) [reply package.json:30]: I don't quite follow the statement about the decorators, since it was only the `GraphQLLookAheadRelations` decorator that was modified.

Making `enableLookAhead` opt-in is still the default I believe, though I would make it so the enable lookahead was on the parent object, not the relation or at least not just the relation. 

Largely the code refactor keeps all the main points, it just changes the shape of the final object that is passed to the query service. I'm not sure how much more complex the parser is in this case compared to the previous one, though it definitely could be improved upon, the previous code parsed the ResolveInfo directly and transformed it into the output  described here: https://github.com/TriPSs/nestjs-query/pull/70#issuecomment-1364126341 , with the obvious changes to the `fieldByTypeName` => `fields` and the simplification of the `XXXConnection` types and returns `Selection<DTO>`

So the output would look like the following (for `todoItem` which is the root):

```
{
  "$args": {
    "id": 1
  },
  "tags": {
    "$args": {
      "sorting": [
        {
          "field": "id",
          "direction": "ASC"
        }
      ]
    },
    "id": true,
    "name": true
  }
}
```

This allows for accessing the properties in a typed manner, allowing previous object to be accessed in the query service like so:

```
const info: Selection<TodoItem> = buildSelectionInfo<TodoItem>( ...some gql resolve info ) // <== actual code to build the obj

info.$args /// this gets the args for the object 

// In the case where we want to lookahead for "tag" objects
// we can do something like this:
if (info.tags) {  // fetch tag props  }
// Or for a particular property
if (info.tags.id && info.tags.name) {  
 // do some other look up or extra calculation in the parent relation
 const tag = tagService.getById( info.$args.id ).map( t => {
return { id: t.id . name: t.name } } )

```

Comment (TriPSs) [reply package.json:30]: What I meant with the decorators was that `@GraphQLLookAheadRelations` now returns almost the same as `@GraphQLResolveInfo`, my idea was that `GraphQLLookAheadRelations` only returns a simple object that contains the main relations and there query. `GraphQLResolveInfo` could then be used if you wanted the full output.

Regarding the outputs, I partly like what you did: 
1. I find the `$args` a bit weird
2. I don't like the double parsing (the lib and then on top of that ours), would prefer we do it completely ourself so we are also not depended on that.

Is it an idea that we keep `GraphQLLookAheadRelations` as is and change the ouput of `GraphQLResolveInfo` to more match yours?

Comment (mrvinceit) [reply package.json:30]: Oh ok, I needed to look back and review previous comments https://github.com/TriPSs/nestjs-query/issues/55#issuecomment-1364111247.

On points 1 & 2:
1. Regarding the `$args` weirdness, there's also a similar `$aliases` prop too 😆. But the idea here was to minimize conflicting properties that a DTO might have since the resulting `SelectionInfo<DTO>` object has the field properties directly on it. 

2. Also, I definitely agree with you on the double parsing, I was running into some typing issues previously when trying to do a single parsing, and left a note-to-self/all about that issue https://github.com/TriPSs/nestjs-query/pull/71/files#diff-a8c4f481736b2c380dd5f16a492f5efd1da1e0e7a5f9f82d3408f1a1de71479fR106. But I did want to get this draft PR reviewed and discussed before jumping back in. 

Thinking about it further, it seems there is very little need for the `GraphQLResolveInfo` decorator, it's a bit redundant since there is already the `@Info` provided by the `nestjs/graphql` package. The fact that `GraphQLResolveInfo` decorator returns two **vastly** different objects based on a single parameter also doesn't sit well with me especially since decorators should primarily be simple, with a single responsibility, blah, blah, blah.

Basically, someone who is creating a resolver and needs the resolver info can get it and do whatever they want already.

`@GraphQLLookAheadRelations` what is the use case that the proposed `SelectionInfo` change does not solve, or does not solve well, also assuming that we get rid of the `@GraphQLResolveInfo` decorator?

Comment (TriPSs) [reply package.json:30]: > `@GraphQLLookAheadRelations` what is the use case that the proposed `SelectionInfo` change does not solve, or does not solve well, also assuming that we get rid of the `@GraphQLResolveInfo` decorator?

The way it's implemented here it returns a lot more info then needed, it could / needs only to return the relations that are needed/can be looked ahead, therefore I thought it was better to have two decorators: 
1. Only return the relations that are used and have look ahead enabled, with there args
    1. Prefer to keep this one as simple as possible so that when you only use the query service directly you don't need to build a complete object but just a simple array with the relation and possible the query)
3. Decorator to fetch the completed query object if you want to do more based on other args/fields

If the second one is not needed, we could remove it but I think it's also nice to provide it as it also makes the creation of the first decorator easier.

Regarding the `$args` weirdness: ~~I can understand that the DTO may have `args` but since all fields will be inside `fields` does that really matter?~~ I see all props are directly on the object instead of fields.

I'm also trying to see the major differences between your implementation and mine, expect that `fields` is removed and it's directly on object. Would it not be preferred to keep the `fields` so we can better type it? And you won't need to check if the field is a boolean or object. Since this interface does type everything correctly:
```ts
type QueryResolveFields<DTO> = {
  [key in keyof DTO]: QueryResolveTree<
    // If the key is a array get the type of the array
    DTO[key] extends ArrayLike<unknown> ? DTO[key][number] : DTO[key]
  >
}

export interface QueryResolveTree<DTO> {
  name: string
  alias: string
  args?: Query<DTO>
  fields: QueryResolveFields<DTO>
}
```

Comment (mrvinceit) [reply package.json:30]: Ok cool, I think we are in general agreement that `@GraphQLResolveInfo` is not absolutely necessary, so let's just agree to remove it.

The way relations work (at least to me) in query service (`queryRelation`, `findRelation`, `countRelation` ) is that they are handled per relation, and therefore the `SelectRelations` parameter would be for all sub-relations of that relation, not the relation itself. 

> The way it's implemented here it returns a lot more info then needed, it could / needs only to return the relations that are needed/can be looked ahead, therefore I thought it was better to have two decorators:
> 

If you have an object that has four relations and all you care about is "looking ahead" for one of them, then sure you could consider the rest of the information unnecessary, but the impact is conceivably negligible:
- Regardless of the number of relations with `enableLookAhead` you still need to do the work of parsing the **_entire_** ResolveInfo object, because of nested objects and the like 
- Accessing the relation is still a constant operation, since the relation is indexed on the prop name, not something like a number or other id:
```javascript
// Ex: Having an object with 4 relations ([Relation1, Relation2, Relation3, Relation4])
// so having extra relations (or properties) makes no difference in usage

//if I want "Relation1"
 obj.relation1 // or obj.relation2 , etc. etc.
```

Regarding `QueryResolveTree` vs `SelectionInfo`, so highlighted differences are:
- `name` is not needed since it is already provided for by the property name. I did toy with the helper prop [`$name`](https://github.com/TriPSs/nestjs-query/compare/feature/look-ahead...mrvinceit:nestjs-query:feature/look-ahead#diff-40a14a114285e0a1e8250c15479219daa6bd0dffaf210a30a105efdf4aa753c1R28), and decided against it but I can be persuaded 
- `QueryResolveTree` doesn't appear to play well with aliases types, `SelectionInfo` handles aliasing by making aliased props explicit and accessible via $aliases prop, which makes the rest of the object graph easier to reason about IMO
- `QueryResolveTree` args is typed as Query<T>, this is incorrect for the args type this should be a more open type (admittedly improving the initial parsing can create closer types for the object if it is a connection type, complex object, etc)
- Exclusion of `fields` is a matter of preference since it removes a level of indirection, there isn't an improvement in typing here, nor does it avoid the need to check if the query requested a certain field or not. You would still have to do something like this:
```javascript
if (obj.fields.id) {
 // do something with obj.fields.id as a QueryResolveTree object
}
```

Comment (TriPSs) [reply package.json:30]: I think we are getting there 😄 

> If you have an object that has four relations and all you care about is "looking ahead" for one of them, then sure you could consider the rest of the information unnecessary, but the impact is conceivably negligible:

Don't understand this one? If you where to enable the look ahead for all four of those relations the array will contain all four of them?

> `QueryResolveTree` args is typed as Query<T>, this is incorrect for the args type this should be a more open type (admittedly improving the initial parsing can create closer types for the object if it is a connection type, complex object, etc)

This could indeed be changed to accept a wider type, main reason it is `Query<T>`is because mainly when it's generated it will be/accept the `Query<T>` type.

To go further on the changes in the PR, I see that `createSelectionInfo` inside the `GraphQLLookAheadRelations` is also not using the relations that are fetched above that have the look ahead enabled, meaning that now all possible relations will be looked ahead.

Was also looking a bit more on the return type of `GraphQLLookAheadRelations` and how it is passed down, could we do something to make it easier for the `typeorm-query.serivce.ts` to read and use? So, mainly not using the same interface for the reason that the query services now don't need to include anything (or know anything) of the GraphQL package.

File: packages/query-graphql/src/decorators/graphql-look-ahead-relations.decorator.ts [hunk 2]
```diff
@@ -35,8 +35,8 @@ export const GraphQLLookAheadRelations = <DTO>(DTOClass: Class<DTO>): ParameterD
     return () => []
   }
 
-  return createParamDecorator((data: unknown, ctx: ExecutionContext): SelectRelation<DTO>[] => {
+  return createParamDecorator((data: unknown, ctx: ExecutionContext): SelectionInfoType<DTO> => {
     const info = GqlExecutionContext.create(ctx).getInfo<ResolveInfo>()
-    return createLookAheadInfo<DTO>(relations, simplifyResolveInfo(info))
+    return createSelectionInfo(info)
```
Comment (TriPSs) [inline packages/query-graphql/src/decorators/graphql-look-ahead-relations.decorator.ts:40]: Could this not become: 
```
    return createLookAheadInfo<DTO>(relations, createSelectionInfo(info))
```

That we than also update the `createLookAheadInfo` to work with that object.

File: package.json (hunk 1 above)
Comment (TriPSs) [reply package.json:30]: Still thinking of keeping the two decorators though, one returning the existing `SelectRelation<DTO>[]` (Or a more similar object like you have but then only with the relations that have look ahead enabled)

And a second one with the full response of `SelectionInfo`, mainly because I can see cases where I could like to have the full info, but since the query services only want the relations to load we can use `GraphQLLookAheadRelations` more as internal one.

File: packages/query-graphql/src/decorators/graphql-look-ahead-relations.decorator.ts (hunk 2 above)
Comment (mrvinceit) [reply packages/query-graphql/src/decorators/graphql-look-ahead-relations.decorator.ts:40]: Sure that's possible

File: package.json (hunk 1 above)
Comment (mrvinceit) [reply package.json:30]: I'm actually not really clear on the purpose that the `SelectInfo` or `SelectRelation` serves for the TypeORM implementations and the like, I kind of assumed it was doing something appropriately similar via the query builder. 

So advantage does `SelectRelation` have for these implementations?

Comment (TriPSs) [reply package.json:30]: `SelectRelation` is meant to be passed down so the query services can query correctly, in case of TypeORM do a `leftJoinAndSelect` instead of a `leftJoin` so that the data is fetched immediately and there will be no need for the data loader to do a separate query,

Comment (TriPSs) [reply package.json:30]: > I think that where we are looking at things separately, is that for the original ask, the issue was never really an issue with relations, which I felt was largely covered, but that the relation could contain a requested field that needed args and I couldn't get those args from what was passed to the the xxxRelations methods in the query service.

I see, this is I think were we mainly mis match because I think the query service itself should not need to know anything related to the query, this way it's easier to use the query service outside the GraphQL query. 

Addition to that (and the two decorators) my idea was to create two:
1. One that will only check which relations to look ahead and pass these down to the service, so the service can make sure that no additional query needs to be done and the data will already be on the entity
2. Second one that will return the complete parsed query (the `SelectInfo`), if the user than needs to do something based on args it can get it there, parse it and do based on that the corrected query on the query service.

How I understand you now you would like to already have the complete `SelectInfo` inside the query service instead of fetching it inside the resolver easily right?

Comment (mrvinceit) [reply package.json:30]: > This could indeed be changed to accept a wider type, main reason it is `Query<T>`is because mainly when it's generated it will be/accept the `Query<T>` type.
> 

I believe I'm more strict about this since the `Query<T>` is only _really_ appropriate for `ConnectionTypes` and based on the `QueryResolveTree` all properties would have this arg type which doesn't make sense.

> To go further on the changes in the PR, I see that `createSelectionInfo` inside the `GraphQLLookAheadRelations` is also not using the relations that are fetched above that have the look ahead enabled, meaning that now all possible relations will be looked ahead.
~Yeah this was on purpose, since this is a draft PR I left a lot of original code in place, ~

> meaning that now all possible relations will be looked ahead.
What do you really mean by "looked ahead"? 
For example, is your concern that a DB implementation will now query all the relation tables because of the code change? Or do you just mean that we are parsing the ResolveInfo for these relations?

If we're worried about the former being the case then makes more sense to me than if we are worried about the latter

Comment (TriPSs) [reply package.json:30]: Sorry what I meant was that now all relations that are fetched, if they have it enabled or not, will now be joined and selected.

Comment (mrvinceit) [reply package.json:30]: We're definitely getting there! 🤣 

> How I understand you now you would like to already have the complete `SelectInfo` inside the query service instead of fetching it inside the resolver easily right?

Yes exactly! Since this saves us from having to recreate all the resolvers which defeats the point of the `nestjs-query` doesn't. This also makes the query service much easier to use for custom stuff since you have everything you need about a standard GQL query but in typed-DTO form and you can do all your special logics.

> the query service itself should not need to know anything related to the query, this way it's easier to use the query service outside the GraphQL query.

I don't quite follow what you mean here, what do you mean by "query" in this context?

> Addition to that (and the two decorators) my idea was to create two:
> 
> 1. One that will only check which relations to look ahead and pass these down to the service, so the service can make sure that no additional query needs to be done and the data will already be on the entity
> 2. Second one that will return the complete parsed query (the `SelectInfo`), if the user than needs to do something based on args it can get it there, parse it and do based on that the corrected query on the query service.

Regarding 1 & 2:
1. Based on what i'm understanding, I think that it is possible to have the full `SelectionInfo` object passed to the query service, but have a `enableLookAhead` as a hint for the `@QueryOptions` or `@Relation`  decorators and in the default implementation of the query services (TypeORM, etc) you can access the metadata to constuct your list of relations to "lookahead". Anyone who is overriding the query services can use the full `SelectInfo` to do what they need based on their business rules anyway so only exposing the enabled doesn't really mean much.

2. I have nothing, this is correct 😃

Comment (TriPSs) [reply package.json:30]: > I don't quite follow what you mean here, what do you mean by "query" in this context?

I mean the GraphQL query.

> Regarding 1 & 2:
> 
> 1. Based on what i'm understanding, I think that it is possible to have the full `SelectionInfo` object passed to the query service, but have a `enableLookAhead` as a hint for the `@QueryOptions` or `@Relation`  decorators and in the default implementation of the query services (TypeORM, etc) you can access the metadata to constuct your list of relations to "lookahead". Anyone who is overriding the query services can use the full `SelectInfo` to do what they need based on their business rules anyway so only exposing the enabled doesn't really mean much.
> 2. I have nothing, this is correct 😃

Okay I now see what you mean, I agree that overwriting the resolver then would defeat the purpose a bit of this lib. The only thing that I'm still thinking about is the part above (let me clarify more): the query services now don't know anything about the GraphQL queries and that makes them also work great outside that context (I reuse my query services a lot within each other for example). 

The main thing i'm a bit afraid of is that when we start using more info of graphQL queries inside the query services we will eventually end up with query services that can't be used anymore without the graphQL query.

Maybe we can also think of a way to tackle this different, how do you see that?

Comment (mrvinceit) [reply package.json:30]: > The main thing i'm a bit afraid of is that when we start using more info of graphQL queries inside the query services we will eventually end up with query services that can't be used anymore without the graphQL query.

I don't see that as a problem here as in I don't see anything **_only_** GraphQL applicable at least not yet. I could easily see the query services being useful for a SQL query for example.

Comment (mrvinceit) [reply package.json:30]: > the query services now don't know anything about the GraphQL queries and that makes them also work great outside that context (I reuse my query services a lot within each other for example).

I think this is actually a good reason to begin and stop here. Including the selected fields in this typed-way that is representative of the `Entity` / `DTO` whatever, removes the **Graphql**-ness of it all, since we can represent the entire and exact context of what we are looking to retrieve all at once, before we largely at the mercy of how Graphql does things, mainly `ResolveFields`

Comment (mrvinceit) [reply package.json:30]: Maybe in re-reading your comments, that was what you are worried about (the ability for query service methods to be abused now)?

Comment (TriPSs) [reply package.json:30]: > I don't see that as a problem here as in I don't see anything **_only_** GraphQL applicable at least not yet. I could easily see the query services being useful for a SQL query for example.

Well the query service now only accepts the `Query<T>` which builds the sql in its turn, the underlaying repo also still available as `service.repo` for more raw like queries, so i'm my eyes (and I thought the general idea of the query service) was to accept a query (the `Query<T>`) and build a database query with it. Adding here relations that needs to be fetched with the query is more inline with the purpose of the service.

Personally I think we are going to lose that if we are going to pass the processed GraphQL query to it, where by default the service will not do anything with. 

>I think this is actually a good reason to begin and stop here. Including the selected fields in this typed-way that is representative of the Entity / DTO whatever, removes the Graphql-ness of it all, since we can represent the entire and exact context of what we are looking to retrieve all at once, before we largely at the mercy of how Graphql does things, mainly ResolveFields

That are good points.

While I was typing the above I thought maybe of two possible ways to do this:

----
#### Option 1
We do it as is, we pass the `SelectionInfo` to service directly

#### Option 2
We could create a service like the `AssemblerClass` class, which will wrap it and default just parse the query to see which relations need to be "looked ahead" and call the query service with it. 

The class will receive full info of the query and can then modify what is being called to the query service.

This will still require to create a separate class and overwrite it, but you then don't need to overwrite the resolver what I think is almost never really preferred.

----

What do you think of the options?

Comment (mrvinceit) [reply package.json:30]: > Well the query service now only accepts the `Query<T>` which builds the sql in its turn, the underlaying repo also still available as `service.repo` for more raw like queries, so i'm my eyes (and I thought the general idea of the query service) was to accept a query (the `Query<T>`) and build a database query with it. Adding here relations that needs to be fetched with the query is more inline with the purpose of the service.

I'm absolutely in favor of this, in fact, I suggested this early-on https://github.com/TriPSs/nestjs-query/issues/55#issuecomment-1280888719 🤣🤣🤣.  But before I get too excited, I'm sure there will be some gotch somewhere. But yeah keeping the essentially the `WHERE`, `FROM` `ORDER BY`, and now `SELECT` altogether in the `Query<T>` makes a lot of sense. 



> #### Option 1
> We do it as is, we pass the `SelectionInfo` to service directly
> 

Sure this works, and we can extend the query service with the extra param or extend `Query<T>` with `SelectionInfo`

> #### Option 2
> We could create a service like the `AssemblerClass` class, which will wrap it and default just parse the query to see which relations need to be "looked ahead" and call the query service with it.
> 
> The class will receive full info of the query and can then modify what is being called to the query service.
> 
> This will still require to create a separate class and overwrite it, but you then don't need to overwrite the resolver what I think is almost never really preferred.

Do you mind explaining this more?

Comment (TriPSs) [reply package.json:30]: > Do you mind explaining this more?

Sure, so the `AssemblerClass` transforms DTO's to entities and entities to DTO's, based on that idea I thought we could create a similar service to that (Let's call it `GraphQLQueryService` for now). This service can then turn the provided `SelectionInfo` into a better/other `Query<T>` or even completely overwrite the call to `this.queryService.query`.

This opens up a couple of possibilities:
1. Every service that need's it get's the same `GraphQLQueryService` which will just call a custom methods on the query service with the `SelectionInfo`
2. Overwrite the `GraphQLQueryService` instead of query services and do everything you wanted there with the full graphql info

Additionally to point 2: Now you can just use the query service to only query the database and maybe provide relations it needs to join and select.

Comment (mrvinceit) [reply package.json:30]: > This service can then turn the provided `SelectionInfo` into a better/other `Query<T>` or even completely overwrite the call to `this.queryService.query`.

I'm not sure that this is the way to go here. It may be making more work for ourselves than it is really saving.
I'll use a project that I'm using lib for as an example, it largely uses the TypeORM query service for one the project DTOs, and in this queryservice, some of the queries needs more of the `SelectionInfo` to connect to some REST APIs etc, etc. 

How does the GraphqlQueryService come into play?

Comment (mrvinceit) [reply package.json:30]: Worst case, it would seem that I would need to write two queryservice for a DTO , one for the AssemblyService another for the query service

Comment (TriPSs) [reply package.json:30]: > How does the GraphqlQueryService come into play?

My idea was that it would wrap the query services so the query services could stay as is, but after some thinking I agree that for now the best way would be to add it as a optional value to `Query<T>` and `FindByIdOptions<DTO>`, if, later, we need more we can still do it as a separate service.

**Comment by mrvinceit** [comment]
@TriPSs , So I updated this original draft based on the discussions that have taken place previously.

Key notes:
- Fixed bugs in graph and added support for queries with Fragments and Union types
- Removed secondary parse with `graphql-parse-resolve-info` package, ResolveInfo is now parsed within the nestjs-query
- Remove type contraint on BaseServiceResolver to fix TS2322 issue
- Cast assembler.convertQuery result as unknown to mask typeconversion mismatch

**Comment by TriPSs** [comment]
Initially looking through the changes (on mobile) it's not really how I had it in mind though.

Will update the looking ahead of total count tonight so you can see how I though on tackling it OK?

We can than discuss from there

**Comment by TriPSs** [comment]
Not sure why this PR is now closed but with #151 merged you should have all the info needed, all the info of that decorator is still marked as internal with the warning that it can change so we can still bring changes to that if we want.
